lib/wibox/container/conditional.lua

```diff
@@ -0,0 +1,362 @@
+--    * The old template widget instance
+--
+function conditional:eval()
+
+    -- Avoid called eval multiple timer per event loop cycle due to the
+    -- attached signals. This creates and destroy widgets for nothing
+    self._private.exec_once = false
+
+    local old_selected, selected = self._private.selected ,nil
+
+    for _, v in ipairs(self._private.templates or {}) do
+        if v.when then
+            if v.when(self._private.argument_callback(self)) then
+                selected = v
+                break
+            end
+        elseif self._private.default_when then
+            local ret = self._private.default_when(v, self._private.argument_callback(self))
+
+            -- The function must return something or is considered broken
+            assert(type(ret) == "boolean", "The conditional layout `default_when`"..
+                " must return a boolean"
+            )
+
+            if ret then
+                selected = v
+                break
+            end
+        else
+            assert(false, "A conditional template has neither a `when` property"..
+                " nor a default_when"
+            )
+        end
+    end
+
+    -- Nothing changed, move on
+    if selected == self._private.selected then
+        if selected and selected.updated_callback then
+            selected.updated_callback(self._private.argument_callback())
+        end
+
+        return
+    end
+
+    if self._private.selected and self._private.selected.deselected_callback then
+        self._private.selected:deselected_callback()
+    end
+
+    self._private.selected = selected
+
+    if self._private.selected and self._private.selected.selected_callback then
+        self._private.selected:selected_callback()
+    end
+
+    local old_widget = self._private.widget
+
+    -- Create the widget unless it's already one.
+    self._private.widget = selected and
+        base.make_widget_from_value(selected) or nil
+
+    self:emit_signal("property::selected_template",
+        selected,
+        old_selected,
+        self._private.widget,
+        old_widget
+    )
+
+    self:emit_signal("widget::layout_changed")
+    self:emit_signal("widget::redraw_needed")
+end
+
+--- Add a signal source to automatically call `eval()`.
+--
+-- @DOC_wibox_container_conditional_object_signals_EXAMPLE@
+--
+-- @tparam gears.object object The source object.
+-- @tparam string signal The signal name.
+function conditional:attach_signal(object, signal)
+    table.insert(self._private.attached_signal, {object, signal})
+    object:connect_signal(signal, self._private.eval_handler)
+end
+
+--- Remove a signal source to automatically call `eval()`.
+function conditional:detach_signal(object, signal)
+    object:disconnect_signal(signal, self._private.eval_handler)
+    for k, v in ipairs(self._private.attached_signal) do
+        if v[1] == object and v[2] == signal then
+            table.remove(self._private.attached_signal, k)
+        end
+    end
+end
```

Comment:
I'd move this code closer to the global signal stuff.

Reply:
I will rework this in the next iteration once these fixes are approved (the `gears.connection` stuff mentioned above)